### PR TITLE
Nibble aligned encoding scheme implementation

### LINKED ISSUE #3: Major Encoding Change Roadmap
There are a number of changes planned from main that will change the encoding. This issue organizes the todo list to make the migration happen.

## Current Compression Scheme

The initial compression scheme implementation included:
1. A VLQ encoding of the full value for each column for the first row
2. A VLQ encoding of the delta from the first row to the second row
3. For each subsequent row, a unary encoding header indicating the number of following bits and the bits for the delta-delta from the previous delta.
    1. header 0, 0 bits
    1. header 10, 4 bits
    1. header 110, 7 bits
    1. header 1110, 9 bits
    1. header 11110, 12 bits
    1. header 111110, 15 bits
    1. header 1111110, 18 bits
    1. header 11111110, 32 bits
    1. header 111111110, 64 bits

## Future Compression Scheme

The future compression scheme will include a single bit before each delta-delta to indicate:
* the following is a truncated binary encoding header indicating the number of following bits and the bits for the delta-delta from the previous delta. Each delta-delta is zigzag encoded
    1. 0, 00, 1 bit
    1. 0, 01, 5 bits
    1. 0, 10, 9 bits
    1. 0, 110, 16 bits
    1. 0, 111, 64 bits

* the following is a truncated binary encoding header indicating the number of bit-packed deltas (not delta-deltas) in the next 32-bits. Each delta is zigzag encoded
    1. 1, 00, pad 0, 2 samples (16 bits)
    1. 1, 01, pad 000, 3 samples (10 bits)
    1. 1, 10, pad 0, 4 samples (8 bits)
    1. 1, 110, 8 samples (4 bits)
    1. 1, 111, pad 00, 10 samples (3 bits)

In the updated scheme, the final encoding will include a final pass with an entropy coding with the minimum word size as 4 bits. All headers and delta bit sequences are 4 bit aligned, with octets tending towards 0000 for constant slope and 1111 for 10 consecutive data points within +-3. Values in delta zigzag encoding may also include octets of leading 0s.

## TODO List

- [x] Benchmark the current implementation on real world data: https://github.com/qsib-cbie/tsz/pull/2
  - [x] Compare compression ratios of second pass compression (lz4) on tsz encoding (1.6x common on real xyz data)
- [ ] Implement the new nibble encoding for delta-delta compression
  - [ ] Compare compression ratios and compression speed
- [x] Implement the new nibble encoding for delta compression
  - [x] Compare compression ratios and compression speed
- [x] Implement the greedy delta-delta, delta switching encoding
  - [x] Enable choice of only/both delta/delta-delta via macro
- [x] Implement Compressor/Decompressor choice of column-major or row-major
  - [x] Compare compression ratios and compression speed (column-major is great for second pass)
- [ ] Implement migration utility from old encoding to new encoding

### PR DESCRIPTION
**Pull request by trueb2**
This adds nibble-aligned delta encoding behind the CompressV2 proc macro. It is suitable for a second-pass compression algorithm such as LZ4 and is about 2-4x faster and better compression than the original delta-delta implementation. Delta-delta headers are reserved but not fully implemented in this PR.

### CODE REVIEW

File: tsz-compress/config.toml [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# Determines the compression algorithm (delta or delta-delta) based on relative size differences.
+# A higher value increases the likelihood of using both algorithms, while a lower value leads to an early decision.
+COMPRESSION_SIZE_FACTOR = 5
```
Comment (trueb2) [inline tsz-compress/config.toml]: This file should be removed.

File: tsz-compress/src/bin/example_v2.rs [hunk 2]
```diff
@@ -0,0 +1,48 @@
+use tsz_compress::prelude::*;
+
+extern crate alloc;
+
+#[derive(Debug, Copy, Clone, CompressV2, DecompressV2)]
+pub struct TestRow {
+    t: i64,
+    a: i8,
+    b: i16,
+    c: i32,
```
Comment (trueb2) [inline tsz-compress/src/bin/example_v2.rs:11]: Add a tsz delta choice to the example as column d i32 using i32.

File: tsz-compress/src/lib.rs [hunk 3]
```diff
@@ -158,12 +158,12 @@
 //!
 //!
 
-#![cfg_attr(not(test), no_std, no_main)]
+#![cfg_attr(all(not(feature = "std"), not(test)), no_std, no_main)]
```
Comment (trueb2) [inline tsz-compress/src/lib.rs]: I believe that we don't want no_main specified here.

File: tsz-compress/src/v2/decode.rs [hunk 4]
```diff
@@ -0,0 +1,839 @@
+use crate::prelude::*;
+use crate::v2::consts::headers;
+use alloc::vec::Vec;
```
Comment (trueb2) [inline tsz-compress/src/v2/decode.rs:3]: looks like a missing line. let's run fmt and clippy again on latest Rust versions.

File: tsz-compress/src/v2/decode.rs [hunk 5]
```diff
@@ -0,0 +1,839 @@
...
+        } else {
+            let value = unsafe { self.buf.get_unchecked(self.idx) } & 0x0F;
+            self.upper = true;
+            self.idx += 1;
+            Some(value)
+        }
+    }
+}
+
+pub fn read_full_i128(buf: &[u8; 16]) -> i128 {
```
Comment (trueb2) [inline tsz-compress/src/v2/decode.rs:55]: We need to add docs to these public functions.

File: tsz-compress/src/v2/encode.rs [hunk 6]
```diff
@@ -0,0 +1,369 @@
+use core::fmt::Binary;
+use num_traits::PrimInt;
+
+use crate::prelude::*;
+use crate::v2::consts::headers;
+
+use super::halfvec::{HalfVec, HalfWord};
+
+pub trait Bits: PrimInt + Binary {
```
Comment (trueb2) [inline tsz-compress/src/v2/encode.rs:16]: more docs on everything public

File: tsz-compress/src/v2/encode.rs [hunk 7]
```diff
@@ -0,0 +1,369 @@
...
+#[inline(always)]
+unsafe fn push_thirty_two_bits(q: &mut CompressionQueue<10>, buf: &mut HalfVec) {
+    buf.push(HalfWord::Half(headers::THIRTY_TWO_BITS_ONE_SAMPLE));
+    let value = q.pop().unwrap_unchecked() as u32;
+    buf.push(HalfWord::Full(value));
+}
+
+// Todo: How can we implement this?
+// #[inline(always)]
+// unsafe fn push_sixty_four_bits<T: PrimInt + Bits>();
```
Comment (trueb2) [inline tsz-compress/src/v2/encode.rs]: We can reserve a header for this now right?

File: tsz-compress/src/v2/encode.rs [hunk 8]
```diff
@@ -0,0 +1,369 @@
...
+
+fn emit_popped_values<const N: usize>(
+    bitcounts: &[usize; N],
+    values: &[usize; N],
+    out: &mut HalfVec,
+) {
+    for (bits, value) in bitcounts.into_iter().zip(values.into_iter()) {
+        match bits {
+            0 => out.push(HalfWord::Half(0b0000)),
+            // -1 => out.push(HalfWord::Half(0b0001)),
```
Comment (trueb2) [inline tsz-compress/src/v2/encode.rs]: Having this disabled will hit the _ case

File: tsz-compress/src/v2/halfvec.rs [hunk 9]
```diff
@@ -0,0 +1,424 @@
...
+                            // println!("lower nibble: {:b}", value);
+                        }
+                        HalfWord::Byte(value) => {
+                            // Fill the lower nibble with the upper nibble of the value
+                            byte |= value >> 4;
+                            known_append(&mut bytes, byte);
+                            // Use the lower nibble from the value as the upper nibble
+                            byte = (value << 4) as u8;
+                            // We are still on the lower nibble
+                            // println!("lower byte: {:b}", value);
```
Comment (trueb2) [inline tsz-compress/src/v2/halfvec.rs]: we should get rid of commented out printlns

File: tsz-compress/src/v2/queue.rs [hunk 10]
```diff
@@ -0,0 +1,281 @@
...
+
+    #[test]
+    fn can_init() {
+        let queue: CompressionQueue<10> = CompressionQueue::new();
+        assert_eq!(queue.len(), 0);
+        assert_eq!(queue.is_full(), false);
+        assert_eq!(queue.is_empty(), true);
+    }
+
+    // #[test]
```
Comment (trueb2) [inline tsz-compress/src/v2/queue.rs]: Should these tests be disabled? Some of this functionality looks like it does not need to be tested.

File: tsz-macro/Cargo.toml [hunk 11]
```diff
@@ -14,5 +14,7 @@ version = { workspace = true }
 proc-macro = true
 
 [dependencies]
-quote = "1.0.26"
-syn = { version = "2.0.10", features = ["full"] }
+quote = "1.0.33"
+syn = { version = "2.0.38", features = ["full", "extra-traits"] }
```
Comment (trueb2) [inline tsz-macro/Cargo.toml]: I don't think that we need full anymore. I could be wrong but I think a lot of those features are just for debugging or testing

File: tsz-compress/config.toml (hunk 1 above)
Comment (ameykasbe) [reply tsz-compress/config.toml]: File removed in commit [b2668f5](https://github.com/qsib-cbie/tsz/pull/45/commits/b2668f52e3f6b241c3dc7626bd30df1948f66b29).

File: tsz-compress/src/bin/example_v2.rs (hunk 2 above)
Comment (ameykasbe) [reply tsz-compress/src/bin/example_v2.rs:11]: Added delta choice to the example in commit [4d1a7cb](https://github.com/qsib-cbie/tsz/pull/45/commits/4d1a7cbdcf6e58310ee350d85c175f97e12524a4).

File: tsz-compress/src/lib.rs (hunk 3 above)
Comment (ameykasbe) [reply tsz-compress/src/lib.rs]: Removed no_main in commit [1f7aa53](https://github.com/qsib-cbie/tsz/pull/45/commits/1f7aa53d94659fda8f809565be5a704648029938).

File: tsz-compress/src/v2/decode.rs (hunk 4 above)
Comment (ameykasbe) [reply tsz-compress/src/v2/decode.rs:3]: Updated code with the help of fmt and clippy in commits: 
1. [610dc8d](https://github.com/qsib-cbie/tsz/pull/45/commits/610dc8d970dadaf7da7c88ee93ac263d519745c6)
2. [cb89e6b](https://github.com/qsib-cbie/tsz/pull/45/commits/cb89e6bd8a05faa18638dace6499add750bc4288)
3. [2266965](https://github.com/qsib-cbie/tsz/pull/45/commits/22669652e4045ed8a85945a71385ca84c582d8f0).

File: tsz-compress/src/v2/decode.rs (hunk 5 above)
Comment (ameykasbe) [reply tsz-compress/src/v2/decode.rs:55]: Added documentation in commit [171f758](https://github.com/qsib-cbie/tsz/pull/45/commits/171f7585d744cbcee1e44fe42be0b18db6513ba3).

File: tsz-compress/src/v2/encode.rs (hunk 6 above)
Comment (ameykasbe) [reply tsz-compress/src/v2/encode.rs:16]: Added documentation in commit [171f758](https://github.com/qsib-cbie/tsz/pull/45/commits/171f7585d744cbcee1e44fe42be0b18db6513ba3).

Comment (ameykasbe) [reply tsz-compress/src/v2/encode.rs:16]: I will work on it as a separate feature right away.

File: tsz-compress/src/v2/encode.rs (hunk 8 above)
Comment (ameykasbe) [reply tsz-compress/src/v2/encode.rs]: We are pushing the zigzag values of data into the CompressionQueue, so it is not possible to have negative values popped out of a CompressionQueue.
```rust
unsafe fn write<T: Bits + Sized>(&mut self, index: usize, t: T) {
     let (zbits, zcount) = t.zigzag_bits();
     *self.zigzag.get_unchecked_mut(index) = zbits;
     *self.bitcount.get_unchecked_mut(index) = zcount;
}
```

Removed the commented statement in commit [fe81490](https://github.com/qsib-cbie/tsz/pull/45/commits/fe814907d974d63ac0406c6ee9545d8fd7951a6b).

File: tsz-compress/src/v2/halfvec.rs (hunk 9 above)
Comment (ameykasbe) [reply tsz-compress/src/v2/halfvec.rs]: Removed commented out println in commit [71e143f](https://github.com/qsib-cbie/tsz/pull/45/commits/71e143fb79db34b659fbe859773cc39ee311f724).

File: tsz-compress/src/v2/queue.rs (hunk 10 above)
Comment (ameykasbe) [reply tsz-compress/src/v2/queue.rs]: Enabled and updated the unit tests according to CompressionQueue in commit [e90280e](https://github.com/qsib-cbie/tsz/pull/45/commits/e90280ec972c8e5ec2fc98867dcd24b04841be23).

File: tsz-macro/Cargo.toml (hunk 11 above)
Comment (ameykasbe) [reply tsz-macro/Cargo.toml]: Updated syn dependency to have only the default features in commit [58a64f9](https://github.com/qsib-cbie/tsz/pull/45/commits/58a64f929b9609d748c4095d2e11e43df4f9b7e8).

File: tsz-compress/src/v2/encode.rs (hunk 6 above)
Comment (ameykasbe) [reply tsz-compress/src/v2/encode.rs:16]: Updated in commit [d0ad91a](https://github.com/qsib-cbie/tsz/pull/46/commits/d0ad91a5f8c5c2370c990649891cfcc87a9e4915).

**Pull request merged**